Модульное тестирование

zklevsha opened the pull request:
---
:mortar_board: [Модульное тестирование](https://up.htmlacademy.ru/vue/1/user/1964457/tasks/8)

Commit:
Added tests fro ItemCounter.vue

Commit:
Added tests for RadioButton.vue

Commit:
Added tests for SelectorItem.vue

Commit:
Added tests for Builder module

Commit:
Added tests for Cart module

Commit:
Added tests for Orders module

Commit:
Added tests for Profile module

Commit:
Added tests for Cart view

Commit:
added tests for Index, Login, Orders, Profile views

Commit:
Added tests for layouts

Djaler commented on src/frontend/src/common/components/__tests__/ItemCounter.spec.js:
```diff
@@ -0,0 +1,66 @@
+import ItemCounter from "@/common/components/ItemCounter";
+
+describe("ItemCounter with min values", () => {
+  const wrapper = mount(ItemCounter, {
```
не стоит монтировать компонент непосредственно в теле describe. Так у тебя получается, что один и тот же инстанс шарится между всеми тестами в этом describe, из-за чего есть вероятность того, что тесты могут быть неявно связаны между собой, зависеть от порядка выполнения и т.д.
Лучше вынести монтирование в beforeEach, например

Djaler commented on src/frontend/src/layouts/__tests__/AppLayoutDefaultHeader.spec.js:
```diff
@@ -0,0 +1,95 @@
+  it("renders cart price", () => {
+    createComponent({ localVue, store, router });
+    expect(wrapper.find(".header__cart > a").text()).toBe(
+      `${store.getters["Cart/cartPrice"]} ₽`
```
в подобных местах лучше явно задать в тесте ожидаемое значение (`200 ₽`, например), чтобы код теста не повторял код реализации. Просто может быть такое, что код окажется сломан. Например, геттер будет всё время возвращать 0. А тест не упадет, поскольку будет проверять на то же самое значение

Djaler commented on src/frontend/src/modules/builder/components/__tests__/BuilderDoughSelector.spec.js:
```diff
@@ -0,0 +1,47 @@
+    expect(wrapper.find("label").attributes("class")).toBe(
+      "dough__input dough__input--light"
+    );
+  });
```
хорошо бы тут проверять обработку выбора одной из опций

Djaler replied:
ну и аналогичные моменты в других компонентах тоже

Commit:
Added beforeEach/afterEach to common/componetns

Commit:
Stopped using getters in tests

Commit:
BuilderDoughSelector: Added large Dough option and added change option test

Commit:
BuilderIngredientsSelector.vue: added tests for sauce change logic, add/remove ingredient logic

Commit:
BuilderPizzaView.vue: added test for add to cart logic

Commit:
BuilderSizeSelector.vue: added tests for medium size option and test for updating size logic

Commit:
CartAdditionalItemCounter: Added additional tests, moved create_component() to beforeEach()

Commit:
CartAdditionalItems: added tests for increase/decrease methods
Mock Vuex: added clone_deep to all objects during mock Vuex initialization

Commit:
CartPizzaList: added tests for add/remove pizza, fixed bug with wrong mutation name

Commit:
CartAddressForm: moved create_component call to beforeEach()

Commit:
CartPopup: moved createComponent to beforeEach

Commit:
OrdersHeader: added test for repeat order logic

Commit:
Profile:
 - moved createComponent to beforeEach
 - added test on editAddress/deleteAddress event emition

Commit:
layouts: moved createComponent to beforeEach

Commit:
Added /cart url test
